src/graph.cc

```diff
@@ -10,113 +10,410 @@ void Graph::LoadCSC(torch::Tensor indptr, torch::Tensor indices) {
+void Graph::CSC2CSR() {
+  SetCOO(GraphCSC2COO(csc_));
+  int64_t num_rows =
+      (row_ids_.has_value()) ? row_ids_.value().numel() : num_rows_;
+  SetCSR(GraphCOO2CSR(coo_, num_rows));
+}
+
+void Graph::CSR2CSC() {
+  SetCOO(GraphCSR2COO(csr_));
+  int64_t num_cols =
+      (col_ids_.has_value()) ? col_ids_.value().numel() : num_cols_;
+  SetCSC(GraphCOO2CSC(coo_, num_cols));
+}
+
+std::tuple<torch::Tensor, torch::optional<torch::Tensor>,
+           torch::optional<torch::Tensor>>
+Graph::PrepareDataForCompute(int64_t axis) {
+  assertm(axis == 0 || axis == 1, "axis should be 0 or 1");
+  torch::Tensor indptr;
+  torch::optional<torch::Tensor> data, e_ids;
+  data = GetData();
+  if (axis == 0) {
+    if (csc_ == nullptr) {
+      CSR2CSC();
+    }
+    indptr = csc_->indptr;
+    e_ids = csc_->e_ids;
+  } else {
+    if (csr_ == nullptr) {
+      CSC2CSR();
+    }
+    indptr = csr_->indptr;
+    e_ids = csr_->e_ids;
   }
-  return this->csc_->indices;
+  return {indptr, data, e_ids};
 }
 
-/**
- * @brief Returns the set of all nodes of the graph. Nodes in return tensor are
- * sorted in the order of their first occurrence in {col_ids, indices}.
- * For example,
- *    if graph.csc_.col_ids = [0, 2, 4, 2] and graph.csc_.indices = [4, 2, 1],
- *    graph.AllIndices() will be [0, 2, 4, 1]
- *
- * @return torch::Tensor
- */
-torch::Tensor Graph::AllIndices() {
-  if (is_subgraph_) {
-    torch::Tensor cat = torch::cat({csc_->col_ids, csc_->indices});
-    return TensorUnique(cat);
+torch::Tensor Graph::Sum(int64_t axis, int64_t powk) {
+  torch::Tensor indptr, out_data;
+  torch::optional<torch::Tensor> in_data, e_ids;
+  std::tie(indptr, in_data, e_ids) = PrepareDataForCompute(axis);
+  out_data = GraphSum(indptr, e_ids, in_data, powk);
+  return out_data;
+}
+
+torch::Tensor Graph::L2Norm(int64_t axis) {
+  torch::Tensor indptr, out_data;
+  torch::optional<torch::Tensor> in_data, e_ids;
+  std::tie(indptr, in_data, e_ids) = PrepareDataForCompute(axis);
+  out_data = GraphL2Norm(indptr, e_ids, in_data);
+  return out_data;
+}
+
+c10::intrusive_ptr<Graph> Graph::Divide(torch::Tensor divisor, int64_t axis) {
+  auto ret = c10::intrusive_ptr<Graph>(std::unique_ptr<Graph>(
+      new Graph(is_subgraph_, col_ids_, row_ids_, num_cols_, num_rows_)));
+  torch::Tensor indptr, out_data;
+  torch::optional<torch::Tensor> in_data, e_ids;
+  std::tie(indptr, in_data, e_ids) = PrepareDataForCompute(axis);
+  ret->SetCSC(csc_);
+  ret->SetCSR(csr_);
+  out_data = GraphDiv(indptr, e_ids, in_data, divisor);
+  ret->SetData(out_data);
+  return ret;
+}
+
+c10::intrusive_ptr<Graph> Graph::Normalize(int64_t axis) {
+  auto ret = c10::intrusive_ptr<Graph>(std::unique_ptr<Graph>(
+      new Graph(is_subgraph_, col_ids_, row_ids_, num_cols_, num_rows_)));
+  torch::Tensor indptr, out_data;
+  torch::optional<torch::Tensor> in_data, e_ids;
+  std::tie(indptr, in_data, e_ids) = PrepareDataForCompute(axis);
+  ret->SetCSC(csc_);
```

Comment:
The generated graph uses the same graph topology as the original graph. It means when we cannot have any mutation on the topology of a graph. Otherwise, it will also change its generated graph. The good thing is we don't need extra copy for the graph topology for these operators. The risk is we need to create a new graph even with minor changes on the graph. I think the risk is acceptable for use because we conduct operators in batch. So we should make a deal that later operators should also follow this immutable development pattern.

Reply:
Yep I agree with that.